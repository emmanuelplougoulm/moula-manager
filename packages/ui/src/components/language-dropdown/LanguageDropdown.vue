<template>
  <div class="container">
    <div class="text-big-regular select" data-testid="select" @click="toggle">
      <SvgIcon :value="modelValue" class="select-flag" />
      <KeyboardArrowDown class="select-arrow-down" />
    </div>
    <div v-show="opened" class="options">
      <div
        v-for="option in options"
        :key="option.id"
        class="text-big-regular option"
        :data-testid="`option-${option.id}`"
        @click="selectOption(option)"
      >
        <Check v-show="showOptionCheck(option.value)" class="option-check" />
        <div class="option-text">
          {{ option.text }}
        </div>
        <SvgIcon :value="option.value" class="option-flag" />
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import KeyboardArrowDown from '../icons/KeyboardArrowDown.vue';
import Check from '../icons/Check.vue';
import SvgIcon from '../icons/SvgIcon.vue';
import type { DropdownOptionType } from '../../types';

const props = defineProps({
  options: {
    type: Array as () => DropdownOptionType[],
    required: true
  },
  modelValue: {
    type: String,
    required: true
  }
});

const opened = ref(false);

const showOptionCheck = (value: string) => {
  return value === props.modelValue;
};
const toggle = () => {
  opened.value = !opened.value;
};
const emit = defineEmits(['update:model-value']);
const selectOption = (option: DropdownOptionType) => {
  opened.value = false;
  emit('update:model-value', option.value);
};
</script>

<style scoped>
.container {
  position: relative;
}

:is(.select, .option) {
  box-sizing: border-box;
  cursor: pointer;
}

.select {
  display: flex;
  justify-content: flex-end;
  align-items: center;
}

.select-arrow-down {
  color: var(--neutral-30);
  width: 1.5rem;
  height: 1.5rem;
}

.select-flag {
  width: 1.625rem;
  height: 1.625rem;
}

.options {
  color: var(--neutral-20);
  border-radius: 0.5rem;
  box-shadow: var(--box-shadow);
  background-color: var(--neutral-100);
  position: absolute;
  top: 1.625rem;
  right: 0;
  width: 10.75rem;
  z-index: 1;
}

.option {
  padding: 0.75rem 1rem 0.75rem 0.5rem;
  display: grid;
  grid-template-columns: 1.25rem 1fr 1.25rem;
  grid-template-areas: 'icon text flag';
  align-items: center;
  gap: 0.5rem;
  line-height: 1.375rem;
}

.option:not(:first-child) {
  border-top: 0.063rem solid var(--neutral-75);
}

.option-check {
  grid-area: icon;
  color: var(--neutral-20);
  width: 1.25rem;
  height: 1.25rem;
}

.option-text {
  grid-area: text;
}

.option-flag {
  grid-area: flag;
  width: 1.313rem;
  height: 1.313rem;
}
</style>
