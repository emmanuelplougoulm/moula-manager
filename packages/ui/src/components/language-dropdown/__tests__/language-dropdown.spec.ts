import { mount } from '@vue/test-utils';
import LanguageDropdown from '../LanguageDropdown.vue';

const options = [
  {
    id: 0,
    text: 'FranÃ§ais',
    value: 'fr'
  },
  {
    id: 1,
    text: 'English',
    value: 'en'
  }
];

describe('LanguageDropdown component tests', () => {
  it('should display the default option flag icon if no option has been selected yet', () => {
    const wrapper = mount(LanguageDropdown, {
      props: {
        modelValue: 'en',
        options: options
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should emit a update:modelValue event if an option has been selected ', async () => {
    const wrapper = mount(LanguageDropdown, {
      props: {
        modelValue: 'fr',
        options: options
      }
    });
    const option = wrapper.find("[data-testid='option-0']");
    await option.trigger('click');

    expect(wrapper.emitted()).toHaveProperty('update:model-value');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
