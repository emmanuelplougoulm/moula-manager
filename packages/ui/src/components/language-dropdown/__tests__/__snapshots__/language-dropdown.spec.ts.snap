// Vitest Snapshot v1

exports[`LanguageDropdown component tests > should display the default option flag icon if no option has been selected yet 1`] = `
"<div data-v-afe87b12=\\"\\" class=\\"container\\">
  <div data-v-afe87b12=\\"\\" class=\\"text-big-regular select\\" data-testid=\\"select\\">
    <en data-v-afe87b12=\\"\\" alt=\\"en\\" class=\\"select-flag\\"></en>
    <div data-v-b9dd0015=\\"\\" data-v-afe87b12=\\"\\" class=\\"container select-arrow-down\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
        <path d=\\"M480 725.566 226.434 472 283 415.434l197 197 197-197L733.566 472 480 725.566Z\\"></path>
      </svg></div>
  </div>
  <div data-v-afe87b12=\\"\\" class=\\"options\\" style=\\"display: none;\\">
    <div data-v-afe87b12=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-0\\">
      <div data-v-b9dd0015=\\"\\" data-v-afe87b12=\\"\\" class=\\"container option-check\\" style=\\"display: none;\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
        </svg></div>
      <div data-v-afe87b12=\\"\\" class=\\"option-text\\">Français</div>
      <fr data-v-afe87b12=\\"\\" alt=\\"fr\\" class=\\"option-flag\\"></fr>
    </div>
    <div data-v-afe87b12=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-1\\">
      <div data-v-b9dd0015=\\"\\" data-v-afe87b12=\\"\\" class=\\"container option-check\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
        </svg></div>
      <div data-v-afe87b12=\\"\\" class=\\"option-text\\">English</div>
      <en data-v-afe87b12=\\"\\" alt=\\"en\\" class=\\"option-flag\\"></en>
    </div>
  </div>
</div>"
`;

exports[`LanguageDropdown component tests > should emit a update:modelValue event if an option has been selected  1`] = `
"<div data-v-afe87b12=\\"\\" class=\\"container\\">
  <div data-v-afe87b12=\\"\\" class=\\"text-big-regular select\\" data-testid=\\"select\\">
    <fr data-v-afe87b12=\\"\\" alt=\\"fr\\" class=\\"select-flag\\"></fr>
    <div data-v-b9dd0015=\\"\\" data-v-afe87b12=\\"\\" class=\\"container select-arrow-down\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
        <path d=\\"M480 725.566 226.434 472 283 415.434l197 197 197-197L733.566 472 480 725.566Z\\"></path>
      </svg></div>
  </div>
  <div data-v-afe87b12=\\"\\" class=\\"options\\" style=\\"display: none;\\">
    <div data-v-afe87b12=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-0\\">
      <div data-v-b9dd0015=\\"\\" data-v-afe87b12=\\"\\" class=\\"container option-check\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
        </svg></div>
      <div data-v-afe87b12=\\"\\" class=\\"option-text\\">Français</div>
      <fr data-v-afe87b12=\\"\\" alt=\\"fr\\" class=\\"option-flag\\"></fr>
    </div>
    <div data-v-afe87b12=\\"\\" class=\\"text-big-regular option\\" data-testid=\\"option-1\\">
      <div data-v-b9dd0015=\\"\\" data-v-afe87b12=\\"\\" class=\\"container option-check\\" style=\\"display: none;\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
        </svg></div>
      <div data-v-afe87b12=\\"\\" class=\\"option-text\\">English</div>
      <en data-v-afe87b12=\\"\\" alt=\\"en\\" class=\\"option-flag\\"></en>
    </div>
  </div>
</div>"
`;
